bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -184,11 +187,20 @@ private static void tryExec(WorkerExecutionContext ctx) {
                 if (debugEnabled && debug(ctx)) {
                     return;
                 }
-    
+
                 Instruction instruction = ctx.code[ctx.ip];
                 int opcode = instruction.getOpcode();
                 int[] operands = instruction.getOperands();
                 ctx.ip++;
+                if (ctx.interruptible) {
```

Comment:
Yes. We can add a checkpoint to an instruction and can be at any point in the program. As an example when program resume after receiving some input we put a checkpoint to next instruction to ensure data is not lost. So for each instruction, we need to check whether it was checkpointed. Since this is a boolean check, I hope it should be fine for non-interruptible services.

Reply:
Based on the current design, When the execution context waiting on a response, we set a checkpoint on next instruction point to persist in the context with the response ( to ensure response message will be persist after resuming it). But this can be improved by checking the execution context state change. I have created https://github.com/ballerina-platform/ballerina-lang/issues/9853 to improve code by handling only necessary places.

Reply:
https://github.com/warunalakshitha/ballerina-lang/commit/62c267e8ebb0d7563af5d58e3dcab2fceca20494 will remove the check for each and every instruction.